### DISCUSSION TITLE
Fix ECMAScript override mistake

### DISCUSSION THREAD

**Discussion by GulgDev** [endojs/endo#2405] [Ideas]
A lot of packages are unfamiliar with ECMAScript override mistake. So it would be hard to use them with SES. However, I found out that this can be fixed by creating a custom `freeze` function, which accounts for the override mistake:
```js
function freeze(object) {
    for (const [property, descriptor] of Object.entries(Object.getOwnPropertyDescriptors(object)))
        if ("value" in descriptor && descriptor.configurable)
            Object.defineProperty(object, property, {
                get() {
                    return descriptor.value;
                },
                set(value) {
                    if (!Object.hasOwn(this, property))
                        Object.defineProperty(this, property, {
                            __proto__: null,
                            value,
                            writable: true,
                            enumerable: true,
                            configurable: true
                        });
                },
                enumerable: descriptor.enumerable,
                configurable: false
            });
    return Object.seal(object);
}
```
It re-defines every data property to accessor property, whose set function:
1. Checks if the receiver has property with the corresponding name
2. If it doesn't, creates a new data property with the given value

**Comment by erights** [comment]
Hi @GulgDev , thanks for the suggestion.

We do approximately this in `enable-property-overrides.js` in the ses-shim for the set of properties enumerated in `enablements.js`. We chose to do this only for enumerated sets for three levels of severity because 
- that takes care of enough cases for most practical purposes, though we continue to get the occasional rude surprise.
- doing it for all properties would be too high overhead, though this is based on intuition, not measurement.

Please do have a look at those two files. One correction to what you write above: In the case where `Object.hasOwn(this, property)`, you still need to emulate an assignment. For that case, you can emulate the assignment by doing an assignment ;)

See
https://github.com/endojs/endo/blob/59b756eb896c72a75f4c0773f964903364bf7e71/packages/ses/src/enable-property-overrides.js#L108-L110
